### PR TITLE
Add syntactic rule for matching on assignment #716

### PR DESCRIPTION
**Pull request by perok**
Done on the Open Source Spree for flatMap(Oslo) 2018.

### CODE REVIEW

File: scalafix-core/shared/src/main/scala/scalafix/internal/rule/DisableSyntax.scala [hunk 1]
```diff
@@ -90,10 +90,33 @@ final case class DisableSyntax(
...
     def hasNonImplicitParam(d: Defn.Def): Boolean =
       d.paramss.exists(_.exists(_.mods.forall(!_.is[Mod.Implicit])))
 
     val DefaultMatcher: PartialFunction[Tree, Seq[LintMessage]] = {
+      case NoValPatterns(v) if config.noValPatterns =>
+        Seq(
+          errorCategory
+            .copy(id = "noValPatterns")
+            .at(
+              "Pattern matching in val assignment is not guaranteed to be total.",
```
Comment (olafurpg) [inline scalafix-core/shared/src/main/scala/scalafix/internal/rule/DisableSyntax.scala]: > Pattern matching in val assignment can result in match error, use "syntax of code using `x match { ... }` instead".

?

File: scalafix-tests/input/src/main/scala/test/disableSyntax/DisableSyntaxNoValPatterns.scala [hunk 2]
```diff
@@ -0,0 +1,14 @@
+/*
+rules = DisableSyntax
+DisableSyntax.noValPatterns = true
+*/
+package test.disableSyntax
+
+case object DisableSyntaxNoValPatterns {
+  val Right(notFound) = 1.asInstanceOf[Either[String, String]] // assert: DisableSyntax.noValPatterns
```
Comment (olafurpg) [inline scalafix-tests/input/src/main/scala/test/disableSyntax/DisableSyntaxNoValPatterns.scala]: Can we add a test case in the basic DisableSyntax to check that this error is not reported unless `DisableSyntax.noValPatterns = true`?

Comment (perok) [reply scalafix-tests/input/src/main/scala/test/disableSyntax/DisableSyntaxNoValPatterns.scala]: I have added the same line of code in DisableSyntaxBase.

**Comment by olafurpg** [comment]
Great catch with `Defn.Var` 🔥

**Comment by joan38** [comment]
Great one @perok!

**Pull request merged**